### PR TITLE
Relax editor templating

### PR DESCRIPTION
**Pull request by Olical**
I tried to set my editor to `termite -e "vim -f {}"`, termite being a
pretty cool and light terminal I use within my i3wm Arch linux box.

So when I open my editor I want it to launch a terminal with Vim inside
instead of GVim for various reasons.

The validation rejected this at first because it was looking for '{}'
inside ['foo', 'bar', 'baz {}'], essentially. So I need it to look
inside the sub-strings, not just the list.

Then after validation I need to perform the '{}' replacement inside the
sub-string too, not just replacing the whole string.

--------------------

I'm not sure what I'm supposed to do testing wise, I haven't written production Python for quite a few years. Let me know if there's anything you'd like me to do or if this approach is a bad one! Happy to change it, I just would like this sub string thing to work :)

The main issue I has was that you're looking for a standalone '{}' token from shlex, but if you use quotes in your editor command it ends up grouping the '{}' token with other tokens found inside the string.

Thoughts?

[Review on Reviewable](https://reviewable.io/reviews/the-compiler/qutebrowser/1263)

### CODE REVIEW

**Comment by lahwaacz** [comment]
I quess this will fix #1092

**Comment by Olical** [comment]
Hah, @The-Compiler and I ended up with essentially the same solution :) and the same use case with the same terminal. I'm just not sure about the testing, style and side effects. Any advice would be much appreciated.

**Comment by The-Compiler** [comment]
The fail on AppVeyor wasn't your fault - I fixed it in master and hit rebuild now.

**Comment by Olical** [comment]
I rebased off of master (I apologise if that broke any local copies! Should have merged but wasn't sure what you preferred). There's still issues of course to do with linting and tests which I can fix, maybe. I'll give it a go anyway. This is a little different to the Django stuff I was building a few years back.

Some of the failing tests are to do with not expecting "foo{}bar" etc to expand, can I change those so it does expect them to expand? Or maybe they're there for a reason?

**Comment by The-Compiler** [comment]
Sorry for not getting to this earlier! I'm a bit busy with non-qutebrowser stuff recently (and probably will be until mid-February or so... but we'll see!).

I don't think there's an explicit reason for the tests being the way they are - it's just that the tests tested the old behaviour, and the behaviour changed, so you'll need to adjust the tests :wink:

By the way: you can run all checks locally by invoking `tox`, or only the tests by doing `tox -e py35` (or `py34` if you're on Python 3.4). To speed things up you can even do `tox -e py35 -- tests/unit/config` to only run the relevant subset.

File: qutebrowser/misc/editor.py [hunk 1]
```diff
@@ -124,6 +124,6 @@ def edit(self, text):
         self._proc.error.connect(self.on_proc_error)
         editor = config.get('general', 'editor')
         executable = editor[0]
-        args = [self._filename if arg == '{}' else arg for arg in editor[1:]]
+        args = [arg.replace('{}', self._filename) if '{}' in arg else arg for arg in editor[1:]]
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: I think you don't need the `... if '{}' in arg else arg` part - if the arg doesn't contain `{}`, `arg.replace('{}', ...)` will return the unchanged string anyways.

Comment (Olical) [inline qutebrowser/misc/editor.py]: Yep, I thought about that, I was trying to change as little as possible, sounds good though. I'll clean this up when I fix the tests. I'll try to get everything green for you within a week or so.

I know it's a small change but it's the cherry on top for me. I'll be gradually moving myself off of Firefox onto qute, I'll contribute where ever I can if I start using it day to do. Glad it's Python!

Tox sounds cool, I was trying to work that out initially, what tooling I needed and what I was supposed to run. I probably need to get virtualenv working again too.

**Comment by Olical** [comment]
@The-Compiler no rush, but I fixed the travis builds. AppVeyor still looks a little shifty though.

**Comment by The-Compiler** [comment]
Looks like this is due to https://github.com/pypa/virtualenv/issues/858 - I worked around it in ec214f4897afe68bdcded501da32c83fcd073f39 by pinning virtualenv to 14.0.3 for now.

Could you please also add a test for the new behaviour to `tests/unit/config/test_configtypes.py` and `tests/unit/misc/test_editor.py`?

Other than that it looks pretty good to me.

**Comment by Olical** [comment]
I've added a few more tests, if there's any edge cases you want me to try and test please let me know. Is there any nice way to run specific tests with tox? I couldn't work it out so I've kind of been relying on the CI to run the tests.

Running them within i3wm causes windows to flash open and closed very quickly. So running the whole suite locally is a little difficult. I'm guessing there's a nice way to filter.

**Comment by hackebrot** [comment]
Two dashes and you should be able to use **pytest** selection features:

``tox -- tests/test_foobar.py -k blazingfast``

**Comment by The-Compiler** [comment]
Sure - you can run something like `tox -e py35 -- args`, and `args` will be passed to pytest, see [usage and invocation](http://pytest.org/latest/usage.html). So in your case, something like `tox -e py35 -- tests/unit/config/test_configtypes.py tests/unit/misc/test_editor.py` should work.

You can also install Xvfb (sometimes packaged as xserver-xvfb or so), and the windows will automatically pop up in a hidden Xvfb instead.

**Comment by Olical** [comment]
Well damn. Thanks a lot! :)

**Comment by The-Compiler** [comment]
I added another test in 37bb26bdd52e37c994ab43e32f3ea52fa3e1362d and merged this - thanks!

Sorry for the failing checks you saw on Travis/AppVeyor - some things are flaky sometimes, and seems like you were particularly unlucky. Everything was fine with your PR! :smile:

**Comment by Olical** [comment]
Thanks a lot, the test looks good, I'll try to get it correct to begin with
next time :) thanks for the guidance on this one. I can use my terminal Vim
to write things now!

Failing CI is fine though. I write JavaScript for a living, I'm used to
restarting things until they work.

**Comment by lamarpavel** [comment]
Working perfectly, thank you very much!
